### PR TITLE
Added ability to select between different SPH smoothing kernels

### PR DESCRIPTION
**Pull request by j-davies-astro**
I've added an extra option to the sph routines that allows the user to pick a different kernel for making images. New kernels need to be defined as functions in the Kernel() class, and so far I've only added 'spline' (the current pynbody default) and 'wendlandC2', which is the kernel EAGLE uses. If no kernel is specified, 'spline' is used.

### CODE REVIEW

File: pynbody/sph/__init__.py [hunk 1]
```diff
@@ -204,7 +206,23 @@ def get_samples(self, dtype=np.float32):
 
     def get_value(self, d, h=1):
         """Get the value of the kernel for a given smoothing length."""
-        # Default : spline kernel
+
+        if self.type == 'spline':
+            # Default : spline kernel
+            return self.spline(d,h)
+
+        elif self.type == 'wendlandC2':
```
Comment (apontzen) [inline pynbody/sph/__init__.py:214]: This should be implemented as a subclass, not as a string option

File: pynbody/sph/__init__.py [hunk 2]
```diff
@@ -214,13 +232,23 @@ def get_value(self, d, h=1):
...
+        else:
+            f = 0
+
+        return (21. * f) / (16. * math.pi * h ** 3)
+
 
 class Kernel2D(Kernel):
 
-    def __init__(self, k_orig=Kernel()):
+    def __init__(self, type='spline'):
```
Comment (apontzen) [inline pynbody/sph/__init__.py:248]: The parameter passed here should indeed be a `Kernel` object, and an appropriate subclass can then be passed if one wants to use a Wendland kernel

File: pynbody/plot/sph.py [hunk 3]
```diff
@@ -364,7 +368,7 @@ def image(sim, qty='rho', width="10 kpc", resolution=500, units=None, log=True,
 
     width = float(width)
 
-    kernel = sph.Kernel()
+    kernel = sph.Kernel(type=kernel_type)
```
Comment (apontzen) [inline pynbody/plot/sph.py:371]: Here you can create the appropraite kernel object based on the string passed in (i.e. it's ok for the user-facing bit to have strings)

File: pynbody/plot/sph.py [hunk 4]
```diff
@@ -375,14 +379,14 @@ def image(sim, qty='rho', width="10 kpc", resolution=500, units=None, log=True,
         is_projected = _units_imply_projection(sim, qty, units)
 
     if is_projected:
-        kernel = sph.Kernel2D()
+        kernel = sph.Kernel2D(type=kernel_type)
```
Comment (apontzen) [inline pynbody/plot/sph.py:382]: This line should read `kernel = sph.Kernel2D(kernel)`, to construct a 2D kernel over whatever kernel has been previously selected

File: pynbody/plot/sph.py [hunk 5]
```diff
@@ -375,14 +379,14 @@ def image(sim, qty='rho', width="10 kpc", resolution=500, units=None, log=True,
...
 
     if av_z:
         if isinstance(kernel, sph.Kernel2D):
             raise _units.UnitsException(
                 "Units already imply projected image; can't also average over line-of-sight!")
         else:
-            kernel = sph.Kernel2D()
+            kernel = sph.Kernel2D(type=kernel_type)
```
Comment (apontzen) [inline pynbody/plot/sph.py:389]: see comment above

File: pynbody/sph/__init__.py [hunk 6]
```diff
@@ -181,7 +181,7 @@ def rho(self):
 
 class Kernel(object):
 
-    def __init__(self):
+    def __init__(self,type='spline'):
```
Comment (apontzen) [inline pynbody/sph/__init__.py:184]: This option should not be present, rather a subclass should be implemented for a different kernel type

**Review by apontzen — CHANGES_REQUESTED** [review]
Nice to do this but the change in the construction of Kernels will break some other projects. The Wendland kernel should be implemented as a new subclass i.e.

```python
class WendlendKernel(Kernel):
    def get_value(...):
        ....
```

That is the basic point, some detailed comments inline

File: pynbody/plot/sph.py [hunk 7]
```diff
@@ -364,25 +368,31 @@ def image(sim, qty='rho', width="10 kpc", resolution=500, units=None, log=True,
 
     width = float(width)
 
-    kernel = sph.Kernel()
+    if kernel_type == 'wendlandC2':
+        kernel = sph.WendlandC2Kernel()
+    elif kernel_type == 'spline':
+        kernel = sph.Kernel()
+    else:
+        print('Invalid kernel_type specified (options are "spline" and "wendlandC2"). Defaulting to a cubic spline...')
```
Comment (apontzen) [inline pynbody/plot/sph.py:376]: This should be `raise ValueError` not `print`

**Comment by j-davies-astro** [comment]
Thanks for the feedback, I think the implementation should be correct now. Happy to change the way it's handled if the user requests an invalid kernel - right now it defaults back to spline with a printed warning but if you'd like to put something more fancy in I can make further changes.

File: pynbody/plot/sph.py [hunk 8]
```diff
@@ -339,6 +340,9 @@ def image(sim, qty='rho', width="10 kpc", resolution=500, units=None, log=True,
     *linthresh* (None): if the image has negative and positive values
      and a log scaling is requested, the part between `-linthresh` and
      `linthresh` is shown on a linear scale to avoid divergence at 0
+
+    *kernel_type* ('spline'): SPH kernel to use for smoothing. Defualts to a
```
Comment (apontzen) [inline pynbody/plot/sph.py:344]: typo (Def*ua*lts)

**Review by apontzen — COMMENTED** [review]
Almost there!

File: pynbody/plot/sph.py (hunk 8 above)
Comment (j-davies-astro) [reply pynbody/plot/sph.py]: whoops!

**Comment by apontzen** [comment]
Sorry one last thing - I think it's worth checking this actually works (and continues to work) with a test. Under `sph_image_test` you will find a `test_images` function which you could easily adapt to also check asking for the alternative kernel behaves as expected (ie. presumably it should return a different image)

**Comment by apontzen** [comment]
Sorry I meant the latter - check the image is not only different but also correct. Ie just add one more reference numpy file to test against.

**Review by apontzen — APPROVED** [review]

**Pull request merged**